### PR TITLE
KAFKA-12294: forward auto topic request within envelope on behalf of clients

### PR DESCRIPTION
**Pull request by abbccdda**
For the metadata auto topic creation case, it is favorable to use `Envelope` to wrap the CreateTopicsRequest alongside the original client principal for auditing purpose.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java:107]: It seems useful to validate that the api key and api version from the request header match that in the request object.

File: core/src/main/scala/kafka/server/ForwardingManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ForwardingManager.scala]: Could we use `RequestContext` as the argument instead of `Request`? As far as I can tell, that is all we need. Similarly, we can change `AutoTopicCreationManager.createTopics`.

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala]: nit: if we document one parameter, we may as well document the others

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala]: Hmm.. It seems a little bit strange to borrow these fields from the original request, but maybe it's not a bad idea. It would potentially allow us to correlate the create request with the original metadata request. It may be worth a comment in the code of explanation.

also nit: unneeded parenthesis

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala]: I'm trying to come up with a better name. Maybe it's ok to be explicit about the fact that this should be a metadata request. If we change to `Option[RequestContext]`, then maybe we can just call it `metadataRequestContext` or something like that.

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala]: nit: we can probably avoid all of these `get` calls if we add something like this above:
```scala
val channelManager = this.channelManager.getOrElse {
  throw new IllegalStateException("Channel manager must be defined in order to send CreateTopic requests")
}
```

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala]: Don't we need to pass `requestVersion` to `build`?

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala]: This is interesting. I guess the situation is when we have not connected to the controller yet. What happens if the controller does not support this version? It should send back UNSUPPORTED_VERSION in the envelope response I think, right? Should we retry the request or do we just rely on the client to resend the Metadata request so that we can retry?

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala]: nit: this call is shared by both branches, maybe we could simplify a little bit. something like this:
```scala
val request = metadataRequestContext.map { context =>
  ...
}.getOrElse(createTopicRequest)

channelManager.sendRequest(request, requestCompletionHandler)
```

File: core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala]: Do we have any test cases which provide a metadata context?

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (abbccdda) [reply core/src/main/scala/kafka/server/AutoTopicCreationManager.scala]: I guess we could rely on client to retry Metadata request for simplicity.

File: core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala
Comment (abbccdda) [reply core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala]: Added

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala]: This message will probably cause confusion since `ClientResponse` may still indicate an error. How about we add the response to the log message?

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala:194]: Let's add a comment that we rely on the Metadata request to be retried in the case that the latest version is not usable by the controller.

File: core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala]: I think there are few interesting cases that should be covered. This is one of them. We should also have assertions in the successful path that the principal is carried over to the envelope correctly. Another interesting case is when the envelope response indicates an UNSUPPORTED_VERSION error.

File: core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala:246]: Maybe I am missing it, but where is the validation? How do we know that the envelope includes the serialized principal? Maybe we could verify the call to `brokerToController.sendRequest`?

File: core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala]: nit: some helpers would improve readability of these tests. It looks like the 3 new tests share a lot of this initialization logic.

File: core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala
Comment (abbccdda) [reply core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala:246]: I was validating the same principal in principal serde call, see 
```
assertEquals(principal, userPrincipal)

```
will also add a boolean flag to indicate it's being called.

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (dengziming) [inline core/src/main/scala/kafka/server/AutoTopicCreationManager.scala:206]: How about moving `serializeWithHeader` to `RequestContext`, like this:
RequestContext.java
```java
    public ByteBuffer buildRequestEnvelopePayload(ApiKeys apiKeys, short apiVersion, AbstractRequest body) {
        // Borrow client information such as client id and correlation id from the original request,
        // in order to correlate the create request with the original metadata request.
        RequestHeader requestHeader = new RequestHeader(apiKeys,
            apiVersion,
            clientId(),
            correlationId());
        return body.serializeWithHeader(requestHeader);
    }
```
and we can call it like `context.buildRequestEnvelopePayload(ApiKeys.CREATE_TOPICS, requestVersion, createTopicsRequest.build(requestVersion))`.

This trivial change has 2 benefits:
1. Keep the symmetry of `RequestContext` since it already has a `buildResponseEnvelopePayload` method
2. `buildRequestEnvelopePayload` can be reused in the future if we want to forward other client requests to the controller

Please feel free to point my fault if I made any mistakes.

Comment (abbccdda) [reply core/src/main/scala/kafka/server/AutoTopicCreationManager.scala:206]: Thanks for the suggestion, I feel this is a bit overkill as we don't have any specific use case to forward other client requests right now, and feel free to propose a refactoring PR after this one is merged to discuss whether the new approach is better.

File: core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/AutoTopicCreationManagerTest.scala:246]: This ensures that an attempt was made to serialize the principal, but how do we know that it made it to the envelope? I think it would be better to intercept the request that is sent to the channel and verify it directly.

File: core/src/main/scala/kafka/server/AutoTopicCreationManager.scala
Comment (hachikuji) [reply core/src/main/scala/kafka/server/AutoTopicCreationManager.scala:206]: @dengziming It's not a bad idea. We could even simplify it a little since the api key and version can be obtained from the request. I tend to agree that this is kind of a niche usage though, so I'm not sure it calls for the generality. Perhaps you could submit a follow-up once this is merged and we can see what it looks like.

**Review by hachikuji — APPROVED** [review]
LGTM if tests are passing

**Pull request merged**
